### ISSUE TITLE
[Bug]: Unlisted breaking change in v2.6.0 with virtual folders returned by extrenal hooks

### ISSUE THREAD

**Issue by Mathieu-COSYNS** [drakkan/sftpgo#1632]
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration issue.
- [X] This issue is **not** already reported on Github _(I've searched it)_.

### Bug description

In v2.6.0:
- Virtual folder must be created before being returned by the external auth hook (in `user.virtual_folders`).
- External auth hook must include the **virtual folder id**.

If the external auth hook does not provide a virtual folder id SFTPGo will:
- Error with `NOT NULL constraint failed: users_folders_mapping.folder_id` if no virtual folder with the same name exist
- Error with `error creating root directory "/virtual_path" for user "test": mkdir : no such file or directory` when listing the virtual folder via SSH, if a virtual folder with the same name exist
- Work fine in the web UI (???)

Is this new behavior expected ? If the answer is yes, is it possible to add this change in the `Backward incompatible changes` list and create an error clearly state that user external auth hook need to return virtual folders with ids. I personally would prefer to keep the old behavior. This new behavior will create a dependency both way between my external auth and SFTPGo.

![image](https://github.com/drakkan/sftpgo/assets/64072917/8cdea03f-df0c-4441-ab82-1a01bc9a6390)

### Steps to reproduce

I created a repo with the smallest config to reproduce the issue (<https://github.com/Mathieu-COSYNS/sftpgo-external-auth-virtual-folder-issue>)

TLDR;

1. Create a external hook that return a user config with a virtual folder without `id`
1. Try to login with a user that use the external hook

and/or

1. Create a external hook that return a user config with a virtual folder without `id`
1. Create a virtual folder with the same name that the one returned by the hook
1. Try to login with a user that use the external hook

### Expected behavior

When returning a virtual folder from a external auth hook, SFTPGo creates the virtual folder if no virtual folder with the same name exist.

### SFTPGo version

v2.6.0

### Data provider

any

### Installation method

Community Docker image

### Configuration

Checkout  (<https://github.com/Mathieu-COSYNS/sftpgo-external-auth-virtual-folder-issue>)

### What are you using SFTPGo for?

Private user, home usecase (home backup/VPS)

### Additional info

Relevant code changes from SFTPGo v2.5.6 changed in v2.6.0

<https://github.com/drakkan/sftpgo/blob/a6a92f0d69b8cf5e72e38ffa5e693057aa2dc776/openapi/openapi.yaml#L5872>
<https://github.com/drakkan/sftpgo/blob/a6a92f0d69b8cf5e72e38ffa5e693057aa2dc776/internal/dataprovider/sqlcommon.go#L2339-L2350>
<https://github.com/drakkan/sftpgo/blob/a6a92f0d69b8cf5e72e38ffa5e693057aa2dc776/internal/dataprovider/sqlcommon.go#L2574>

**Comment by drakkan** [comment]
Updated release notes, thanks for noticing. See #1349

I think it should work if you return an existing virtual folder by name.

**Comment by Mathieu-COSYNS** [comment]
Hi, thank you for the quick reply. I didn't knew about #1349. I was hopping to still be able to create virtual folders on the fly when returning user configuration. It's a bit annoying but I will adapt my approach for my external auth hook.

> I think it should work if you return an existing virtual folder by name.

As I was explaining above, it does work for the Web UI but not for SSH (and FTP, WebDAV) for those you need the virtual folder id. This behavior should be consistent.

> Error with error creating root directory "/virtual_path" for user "test": mkdir : no such file or directory when listing the virtual folder via SSH, if a virtual folder with the same name exist

One approach could be to change the user/group create/update API  `virtual_folders` to take only virtual folders `virtual_path`, `quota_size`, `quota_files`, `folder_id` instead of full virtual folders configuration since setting other fields will/should not have any impact.

Should we continue this in another issue or reopen this one ?

**Comment by drakkan** [comment]
I did a quick test to confirm, returning an existing folder name is enough, the id is not required. 

The OpenAPI document should be updated in several places to clarify the fields required for adding/updating and the fields returned in GETs. 
If you want to help us and feel comfortable with our [CLA](https://sftpgo.com/cla.html) feel free to send a PR. The PR should be complete and working to be accepted, we have currently no time for reviewing and iterating or pushing back on PRs (this is often more work than doing the change ourself). Thanks for understanding and for using SFTPGo

**Comment by firatkucuk** [comment]
We also hit this barrier. It seems one way communication needs to be changed as 2 ways communication. So currently we are in between "to create a PR for changing the implementation to one way communication as before" or "implementing this behaviour as is". I am open to create a PR for changing the webhook but I want to be sure before spending time whether it fits with your architectural vision or not.

**Comment by drakkan** [comment]
> We also hit this barrier. It seems one way communication needs to be changed as 2 ways communication. So currently we are in between "to create a PR for changing the implementation to one way communication as before" or "implementing this behaviour as is". I am open to create a PR for changing the webhook but I want to be sure before spending time whether it fits with your architectural vision or not.

We want to have a separation between users, folders and groups. Historically the hook automatically created folders because folders were originally included with users, but this design had several limitations. Now we also have groups and we don't want the hook to automatically create groups as well, we should do that if we create folders.

**Comment by firatkucuk** [comment]
@drakkan So is it okay us to create PR to perform login with one way communication instead two?

**Comment by patschi** [comment]
I'm unfortunately "hit" with this change as well. I've been writing a external auth service for myself to authenticate against my QNAP NAS (QNAP doesn't have native SFTP service, so I wanted to use sftpgo as a replacement but still authenticate against QNAP). I wanted to avoid two-way communication for this.

@firatkucuk Given it hasn't changed in 2.7, may I ask if you found any other workaround, or filed a PR for the same?

**Comment by firatkucuk** [comment]
I have asked; can i make it to one way communication version of it. After quite some time of inactivity I gave up and implemented this way but bidirectional communication is not the best way for sure

**Comment by patschi** [comment]
Understood - thanks.

@drakkan Just wondering - are you fine with a PR reverting to old pre-2.6 behavior? (I'd implement the automatic creation of virtual folders when external-auth returns non-existing virtual folders in its json)

**Comment by drakkan** [comment]
> Understood - thanks.
> 
> [@drakkan](https://github.com/drakkan) Just wondering - are you fine with a PR reverting to old pre-2.6 behavior? (I'd implement the automatic creation of virtual folders when external-auth returns non-existing virtual folders in its json)

I understand that embedding virtual folders within users might be convenient for your use case. I anticipated that this [change](https://github.com/drakkan/sftpgo/commit/0413c0471ce6112dabce29df8d47f5345e5d7248) could lead to some [concerns](https://github.com/drakkan/sftpgo/issues/1349#issuecomment-1647413462). However, we've now adopted a model where virtual folders are treated as separate objects, both logically and at the database level. 

Automatically adding or updating virtual folders when modifying users introduces complexity and makes the code harder to maintain, particularly with the new features that have been introduced and that may be introduced in the future. Users and virtual folders are separate entities that can be used together, but virtual folders are not user-specific and can, for example, also be utilized in the EventManager.

That said, I believe you can  reference virtual folders by name (not by id), and you can also parameterize virtual folders through groups.
